porch/controllers/rootsyncdeployments/pkg/controllers/rootsyncdeployment/controller.go

```diff
@@ -0,0 +1,539 @@
+	if err := r.Get(ctx, req.NamespacedName, &rootsyncdeployment); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	myFinalizerName := "config.porch.kpt.dev/rootsyncdeployments"
+	if rootsyncdeployment.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Update the cache with mapping from rollouts to ClusterTargets. It allows the controller
+		// to determine which Rollouts needs to be reconciled based on an event about a cluster.
+		selector, err := metav1.LabelSelectorAsSelector(rootsyncdeployment.Spec.Targets.Selector)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error converting targets selector to labels.Selector: %v", err)
+		}
+		func() {
+			r.mutex.Lock()
+			defer r.mutex.Unlock()
+			r.clusterTargetCache[req.NamespacedName] = selector
+		}()
+
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(&rootsyncdeployment, myFinalizerName) {
+			controllerutil.AddFinalizer(&rootsyncdeployment, myFinalizerName)
+			if err := r.Update(ctx, &rootsyncdeployment); err != nil {
+				return ctrl.Result{}, fmt.Errorf("error adding finalizer: %w", err)
+			}
+		}
+	} else {
+		// Clean up the cache.
+		func() {
+			r.mutex.Lock()
+			defer r.mutex.Unlock()
+			delete(r.clusterTargetCache, req.NamespacedName)
+		}()
+
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(&rootsyncdeployment, myFinalizerName) {
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(&rootsyncdeployment, myFinalizerName)
+			if err := r.Update(ctx, &rootsyncdeployment); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update %s after delete finalizer: %w", req.Name, err)
+			}
+		}
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	res, err := r.syncRootSyncDeployment(ctx, &rootsyncdeployment)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.updateStatus(ctx, &rootsyncdeployment, res); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *RootSyncDeploymentReconciler) updateStatus(ctx context.Context, rds *v1alpha1.RootSyncDeployment, results []result) error {
+	var clusterRefStatuses []v1alpha1.ClusterRefStatus
+
+	for _, res := range results {
+		clusterRefStatuses = append(clusterRefStatuses, v1alpha1.ClusterRefStatus{
+			ApiVersion: res.apiVersion,
+			Kind:       res.kind,
+			Name:       res.name,
+			Namespace:  res.namespace,
+			Revision:   res.revision,
+			Synced:     res.synced,
+		})
+	}
+
+	if equality.Semantic.DeepEqual(rds.Status.ClusterRefStatuses, clusterRefStatuses) {
+		klog.Infof("Status has not changed, update not needed.")
+		return nil
+	}
+	rds.Status.ClusterRefStatuses = clusterRefStatuses
+	return r.Status().Update(ctx, rds)
+}
+
+type result struct {
+	apiVersion string
```

Comment:
Why make a new type here instead of just using the `ClusterRefStatus` type directly?

Reply:
Hmm that is an excellent question. I've updated the PR to just use `ClusterRefStatus` directly.